src/readwrite_array.jl

```diff
@@ -0,0 +1,39 @@
+# A Union of ITensor-defined types that can be read and
+# written when stored in Arrays
+ITensorHDF5Types = Union{TagSet,Index,ITensor,MPS,MPO}
+
+hdf5_type(::Type{<:Index}) = Index
+hdf5_type(::Type{<:AbstractArray{T}}) where {T} = Array{hdf5_type(T)}
+hdf5_type(::Type{T}) where {T} = T
+hdf5_type(::T) where {T} = hdf5_type(T)
+
+function HDF5.write(
+  parent::Union{HDF5.File,HDF5.Group},
+  name::AbstractString,
+  d::AbstractArray{<:ITensorHDF5Types},
+)
+  g = create_group(parent, name)
+  attributes(g)["type"] = "$(hdf5_type(d))"
+  attributes(g)["version"] = 1
+  write(g, "size", collect(size(d)))
+  for c in CartesianIndices(d)
+    write(g, "$(Tuple(c))", d[c])
+  end
+end
+
+function HDF5.read(
```

Comment:
Maybe we should switch to using this for MPS and MPO in version 2 of the ITensor HDF5 file format for those types.

Reply:
Yes, part of the goal of this design would be to use it in more complicated types like MPS and MPO. That would be a good use of the version number.